fs.go

```diff
@@ -0,0 +1,183 @@
+	entityID := publicKeyToFilename(id)
+	return p.fs.Join(registryDir, registryEntityDir, entityID+statementExt)
+}
+
+// Implements Provider.
+func (p *fsProvider) GetEntity(ctx context.Context, id signature.PublicKey) (*EntityMetadata, error) {
+	f, err := p.fs.Open(p.getEntityPath(id))
+	switch {
+	case err == nil:
+	case os.IsNotExist(err):
+		return nil, ErrNoSuchEntity
+	default:
+		return nil, fmt.Errorf("%w: failed to open entity metadata: %s", ErrCorruptedRegistry, err)
+	}
+	defer f.Close()
+
+	entity := new(EntityMetadata)
+	return entity, entity.Load(id, f)
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) BaseDir() string {
+	return p.baseDir
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) Init() error {
+	paths := []string{
+		registryDir,
+		p.fs.Join(registryDir, registryEntityDir),
+	}
+
+	for _, path := range paths {
+		if _, err := p.fs.Stat(path); !os.IsNotExist(err) {
+			return fmt.Errorf("registry already initialized (or corrupted)")
+		}
+
+		if err := p.fs.MkdirAll(path, 0o755); err != nil {
+			return fmt.Errorf("failed to create path %s: %w", path, err)
+		}
+
+		placeholderPath := p.fs.Join(path, placeholderFilename)
+		f, err := p.fs.OpenFile(placeholderPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o644)
+		if err != nil {
+			return fmt.Errorf("failed to create placeholder %s: %w", placeholderPath, err)
+		}
+		f.Close()
+	}
+
+	return nil
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) Verify() error {
+	// Verify entities.
+	entities, err := p.fs.ReadDir(p.fs.Join(registryDir, registryEntityDir))
+	if err != nil {
+		return fmt.Errorf("failed to read entity directory: %w", err)
+	}
+
+	for _, fi := range entities {
+		if filepath.Ext(fi.Name()) != statementExt {
+			continue
+		}
+
+		var id signature.PublicKey
+		if err = id.UnmarshalHex(strings.TrimSuffix(fi.Name(), statementExt)); err != nil {
+			return fmt.Errorf("entity: bad statement filename '%s': %w", fi.Name(), err)
+		}
+
+		if fi.Size() > MaxStatementSize {
+			return fmt.Errorf("entity: statement too big (size: %d max: %d): %s", fi.Size(), MaxStatementSize, fi.Name())
+		}
+
+		if _, err = p.GetEntity(context.Background(), id); err != nil {
+			return fmt.Errorf("entity: bad statement '%s': %w", fi.Name(), err)
+		}
+	}
+	return nil
+}
+
+// Implements MutableProvider.
+func (p *fsProvider) UpdateEntity(entity *SignedEntityMetadata) error {
+	// Make sure the signed entity is valid before processing it.
+	var inner EntityMetadata
+	if err := entity.Open(&inner); err != nil {
+		return fmt.Errorf("bad signed entity metadata: %w", err)
+	}
+	if err := inner.ValidateBasic(); err != nil {
+		return fmt.Errorf("bad signed entity metadata: %w", err)
+	}
+
+	// Check if the entity already exists. In this case, require that the serial number is bumped.
+	existing, err := p.GetEntity(context.Background(), entity.Signature.PublicKey)
+	switch err {
+	case nil:
+		if inner.Serial <= existing.Serial {
```

Comment:
Maybe this check does not even make sense here in any case, as this will have to be separately checked (and enforced) by each registry anyway. (see also: https://github.com/oasisprotocol/metadata-registry-tools/pull/1#discussion_r473847570)

Reply:
Yeah this check is only there as a local sanity check. We need to add something like `verify --updates <old-registry>` so that the CI can perform verification, will add the subcommand.

Reply:
Added new update verification methods and tests.